gup.8
.TH GUP 8 "25 July 1993"
.\"
.\" Cobbled together by Mark Delany <markd@bushwire.apana.org.au>
.\" who knows virtually zilch about nroff and I bet it shows...
.\"
.SH NAME
gup \- A
.IR G roup
.IR U pdate
.IR P rogram
that accepts commands by mail to edit a newsgroup subscription
file for subsequent use by news systems such as
.I INN
and
.IR C-News .
.SH SYNTAX
.B gup
.RB [ \-hvP ]
.RB \-a
.IR active_path
.RB [ \-d
.IR home_directory ]
.RB [ \-l
.IR log_path ]
.if n .ti +.6i
.RB [ \-m
.IR reply_headers ]
.RB [ \-n
.IR newsgroups_path ]
.if n .ti +.6i
.RB [ \-s
.IR  sites_directory ]
.RB [ \-M
.IR Mail_command ]
.SH DESCRIPTION
.PP
The sole purpose of
.I gup
is to automate the tedious process of editing group
selection patterns defined in the news configuration 
files (eg:
``newsfeeds'' for
.I INN
and ``sys'' for
.IR C-News ).
.PP
.I Gup
is of use to news administrators
who spend an inordinate amount of time
editing their news config files
at the behest of the sites they feed.
In fact,
once
.I gup
is installed,
it is quite likely that manual edits of
your ``newsfeeds'' or ``sys'' file will
become a thing of the past.
.PP
.I Gup
is designed to be installed as a mail-server program
that is fed an inbound mail via stdin.
.I Gup
is usually
invoked from a .forward file. Eg:
.PP
.nf
	"|/.../bin/gup \-options...."
.fi
.PP
Each site has an
entry in the ``config'' file containing password
and mail address details
and a group selection file called ``groups'',
see
.BR CONFIG ,
and
.BR GROUPS
for more details.
.PP
The
news administrator of each
site mails commands to
.IR gup .
There are commands to include and exclude group patterns, list the
current patterns for that site and list the available
newsgroups; see
.BR COMMANDS ,
for more details.
.PP
The results are normally mailed back to the site's configured
administrator.
However under some circumstances, the results are
mailed to
the originator or the local administrator; see
.BR PROCESSING ,
for further details.
.PP
.I Gup
does not directly change the
news system's control files
(eg, ``newsfeeds'' for INN).
Instead a trivial shell script must be run to concatenate all
of the changed ``groups'' files together into an appropriately formatted
file for your particular news system. (One is provided in the source
kit for INN).
.PP
Since each site has to be specifically configured in
.IR gup 's
``config'' file,
access can be restricted to administrator's capable of managing
their own group patterns.
.SH OPTIONS
.PP
Options can appear in any order on the command line. The most important
point to note is that all of the paths and directories defined will
normally be absolute paths unless you are intimately familiar with
the way in which
.I gup
changes directories as it processes a mail (the possible exception
here is the Sites_directory).
.PP
.TP
.BI \-a " active_path"
The path of the active file for your news system.
Before accepting any newsgroup identified in a command,
.I gup
validates the group against the active file. The command is
rejected if no match is found.
.TP
.BI \-d " home_directory"
Defines
.IR gup 's
home directory.
.I Gup
changes to this directory as soon as possible after
starting up. If this option is not present, the current directory
is used.
.I Gup
looks for the ``config'' file in it's home directory.
.TP
.BI \-h
Print out a help message showing the command line options, then
exit.
.TP
.BI \-l " log_path"
A record of all significant requests are written to this file. If
the path is relative, then it will be relative to
.IR gup 's
home directory; see the \-d option).
.I Gup
must be able to write to this file.
If the
.BI \-l
option is not used, then
.I gup
uses stderr. This is useful for testing purposes, but
is unlikely to be of use in a .forward file.
.TP
.BI \-m " reply_headers"
When
.I gup
generates a mail response it only generates the ``To: '' header line.
This option defines the path of
a file that contains other RFC882 conformant 
header lines that are piped to the mail program (see the \-M option).
In fact, if this
file contains a body following the headers, then that will
precede any text generated by
.IR gup .
If this path is not an absolute path, then it will be treated as
relative to
.IR gup 's
home directory (see the \-d option).
.TP
.BI \-M " Mail_command"
.I Gup
pipes the rfc822 headers and the body of the mail to the nominated mail
program. Normally, this is configured when
.I gup
is installed, but it can be over-ridden with this option. The mail
command must be able to determine the recipient addresses from
the rfc822 headers.
.TP
.BI \-n " newsgroups_path"
If present, the newsgroups file is used to try and find a matching
description of newsgroup when listed.
.TP
.B \-P
Do
.I not
prune superfluous patterns from a site's ``groups'' file. Before
writing the updated ``groups'' file,
.I gup
applies a fairly rigorous test to the patterns, pruning
any nonsensical or un-necessary patterns. This pruning process
can be quite CP intensive to the extent that it may have a deleterious
effect on your system - thus the ability to disable it.
.TP
.BI \-s " Sites_directory"
Each site's ``groups'' and ``exclude'' file are located in
a unique directory for each site. These site directories are
located in the directory defined with this option. If this is given
as a relative path then it will be relative to
.IR gup 's
home directory (see the \-d option).
.I Gup
will try and create this directory if it does not exist.
.TP
.B \-v
Print out the version number and various compile-time variables,
then exit.
.SH COMMANDS
.PP
.I Gup
scans the body of the mail for commands. Blank lines are ignored and
any data after the ``#'' character is considered a comment. No
continuation is allowed. Many of the commands accept a pattern
as a parameter.
This pattern is identical to the format of the wildmat()
pattern; see
.B wildmat (3)
).
In fact,
.I Gup
purposely uses the
.B wildmat
routine from INN to ensure that the pattern matching characteristics
are identical.
.PP
Valid commands are:
.TP
.BI site " sitename password"
This
.I must
be the first command in the mail.
.I sitename
and
.I password
must match an entry in the ``config'' file.
Only one
.I site
command is allowed per mail. Aliases: "open" and "host".
.TP
.B quit
This command stops
.I gup
from processing the rest of the mail. This is useful if
your mail User Agent tends to automatically append a signature
file to your mail. Alias: "q".
.TP
.BI include " pattern"
The
.I pattern
is checked against the active file. If it matches at
least one newsgroup, the
.I pattern
is placed at the end of the site's
``group'' file as an
.I include
entry.
Only one
.I pattern
per
.I include
command is allowed. If the pattern matches anything in the
site's exclusion list (see
.BR EXCLUSIONS )
then the include will fail.
Aliases: "+" and "inc".
.TP
.BI exclude " pattern"
The
.I pattern
is checked against the active file. If it matches at
least one newsgroup, the
.I pattern
is placed at the end of the site's
``group'' file as an
.I exclude
entry.
Only one
.I pattern
per
.I exclude
command is allowed.
Aliases: "\-" and "exc".
.TP
.BI poison " pattern"
If the
.I pattern
matches at least one crossposted newsgroup the article will not be accepted
even if allowed by an
.I include
statement.
Only one
.I pattern
per
.I poison
command is allowed.
.TP
.B help
Generate a small help message that
briefly describes each command.
There is an implied quit with the help command so there is no
point in placing commands after the help command.
Alias: "h".
.TP
.B list
list all of the
current
.I include
and
.I exclude
patterns in the sites ``groups'' file.
The
output is in a format suitable for feeding back into
.I gup
at a later stage if need be.
Alias: "l".
.TP
.BI delete " pattern"
Delete all
.IR include ,
.I exclude
and
.I poison
patterns in the site's ``groups'' file
that match the
.IR pattern .
``delete *'' is an effective way
of clearing all current patterns.
.TP
.BI newsgroups " pattern"
This command lists out all available newsgroups from the
active file that match the
.IR pattern .
The list includes the description from the newsgroups file
as well as an indication if the site is currently
subscribed to that
group.
Only one
.I pattern
per
.I newsgroups
command is allowed.
Alias: "news".
.SH PROCESSING
.PP
.I Gup
has a number of processing stages. The initialization
stage consists of changing to the home directory (see the \-d option)
and opening the logfile (see the \-l option). At this time,
.I gup
sets the tentative reply-to mail address to the ``backstop'' mail
address
defined when 
.I gup
was compiled (typically the local news
administrator).
.PP
The next stage consists of scanning the inbound mail, noting
.I interesting
mail headers. The most interesting ones
are "To:" and "Reply\-To:".
When a "To:" header is found it becomes the tentative
reply-to mail address. If a "Reply\-To:" header is found it over-rides
any "To:" address to become the new tentative reply-to mail address.
A few others
are noted and logged to help track changes.
.PP
After all the headers have been processed, the body of the mail
is examined for commands. The first command
.B must
be the
.I site
command. Any other data results
in an error mail sent to the tentative reply-to mail address.
If the 
.I site
command contains a name that matches an entry in
the
``config'' file, then the tentative reply-to mail address
is replaced with
the mail address in the ``config'' file.
.PP
The reason for these contortions with tentative reply-to mail
addresses is simply
to deal with the problem of working out who to send a mail to in 
the event of an error. Ideally they should all go back to the
mail address in the ``config'' file, but that information is not
known for quite a significant part of
.IR gup 's
initial processing.
.PP
Once a valid
.I site
command has been accepted,
.I gup
changes to that site's directory
in Sites_directory (see the \-s option) making the Sites_directory and
site's directory as necessary. The site's directory name is the
same as the site's name. In the absence of the \-s option this
will be:
.nf

	$HOME/sites/$site

.fi
Where $HOME is
.IR gup 's
home directory and $site is the name of the site being processed.
.I Gup
locks the site
then loads the site's current ``groups'' file and any
xclusion list if
present (see
.BR EXCLUSIONS
for more details).
.PP
From this point on
.I gup
accepts any command in any order until either the end of the mail,
a quit command a help command or a serious error during processing.
After all commands have been processed,
.I gup
update's the site's ``groups'' file if
changes have been made.
This update includes pruning any superfluous patterns
(unless the \-P option is used).
.I Gup
writes the new patterns to ``groups.new''. It then

gup.c
/*
 * A triv program to parse a mail and accept simple commands.
 *
 * Date         Who     What
 * -------      ----    ----------------------------------------
 * 19 Jul 93    MarkD   Initial coding.
 * 20 Jul 93    MarkD   Incorporated AHerb's marvellous changes
 *
 * Mail comes in on stdin.
 *
 * Valid commands are:
 *
 *      site            name            passwd
 *      include         <group-pattern>
 *      exclude         <group-pattern>
 *      poison		<group-pattern>
 *      list
 *      newsgroups      <group-pattern>
 *      delete          <group-pattern>
 *      quit
 *
 * Case is irrelevant.
 * group can be a wildmat pattern.
 *
 * If any command has changed the current grouplists for a given site
 * then the list is pushed through prune.c which does a fairly thorough
 * test to ensure that the list makes sense. Anything non-sensical is
 * removed.
 *
 * The delete group is checked against the currently subscribed list
 * rather than the active.
 */

#define	MAIN

#include "gup.h"
#include "rfc822.h"


const char *progname = "gup";	/* GC */

static const char *usage =
"\n\
Usage: %s\t[-hvP] -a active_path [-d home_directory] [-l log_path]\n\
\t\t[-m reply_headers] [-n newsgroups_path] [-s sites_directory]\n\
[-M Mail_command] [-G maxgroups]\n\
\n\
Group Update Program: mail-server for remote newfeed changes.\n\
\n\
-a\tThe news system's 'active' file - used to validate newsgroups\n\
-d\tgup's home directory (default: .)\n\
-h\tPrint usage then exit\n\
-l\tLog file for writing significant events (default: ./log)\n\
-m\tfile containing rfc822 headers to send to the recipient\n\n\
-n\tThe newsgroups file - used for newsgroups descriptions\n\
-M\tResultant mail piped to this command\n\
-P\tDo not prune newsgroup selections (prune is CP intensive)\n\
-s\tLocation of the site directories (default: ./sites)\n\
-v\tPrint version and build options, then exit.\n\
-G\tMaximum number of subscriptions allowed\n\
\n";


typedef struct command_s COMMAND;

struct command_s {
    const char *name;
    int (*handler) (const char **);	/* function to handle it */
    int args;
    int needs_site;
    int sets_quit;
};


static void parse_options(int argc, char **argv);
static void parse_headers(void);
static int parse_commands(void);
static void load_active(LIST * exclusion_list);
static void write_groups(void);
static void print_version(void);
static void log_mail_headers(void);
static void item_print(int indent, int *column, const char *str);

static int getoneline(FILE * fp);
static int tokenize(char *cp, const char **tokens, int max_tokens);
static int command_cmp(const char *str, const char *cmd);
static int site(const char **tokens);
static int include(const char **tokens);
static int exclude(const char **tokens);
static int poison(const char **tokens);
static int insert_group(int not_flag, const char *gname);
static int delete(const char **tokens);
static int list(const char **tokens);
static int quit(const char **tokens);
static int newsgroups(const char **tokens);


/* Here's the command list from the body of the mail */
static COMMAND C_list[] =
{
    {"inc*lude", include, 2, TRUE, FALSE},
    {"+", include, 2, TRUE, FALSE},
    {"exc*lude", exclude, 2, TRUE, FALSE},
    {"poi*son", poison, 2, TRUE, FALSE},
    {"-", exclude, 2, TRUE, FALSE},
    {"l*ist", list, 0, TRUE, FALSE},
    {"he*lp", help, 0, FALSE, TRUE},
    {"del*ete", delete, 2, TRUE, FALSE},
    {"s*ite", site, 3, FALSE, FALSE},
    {"o*pen", site, 3, FALSE, FALSE},
    {"host", site, 3, FALSE, FALSE},
    {"news*groups", newsgroups, 2, TRUE, FALSE},
    {"q*uit", quit, 0, TRUE, TRUE},
    {"--", quit, 0, TRUE, TRUE},
    {"c*lose", quit, 0, TRUE, TRUE},
    {NULL}
};


static int seen_site = FALSE;	/* If site command accepted */
static int do_prune = TRUE;
static char *sitename;

static const char *site_directory = SITE_DIRECTORY;

static char curr_line[MAX_LINE_SIZE];

static int quit_flag;

/* Variables used to control mailed results */
static char *h_from = NULL;
static char *h_reply_to = NULL;
static char *h_date = NULL;
static char *h_subject = NULL;
static char *h_message_id = NULL;
static char *h_return_path = NULL;

/*
 * The major note for this program is that calls to logit() should be
 * avoided as much as possible prior to the 'site' command. The reason
 * for this is that we have a "less-than-ideal" mail recipient before
 * that time.
 *
 * You see the mail recipient starts out as the backstop mailid
 * (typically news or root or some local admin name) as the inbound
 * headers are cracked this changes to the FROM: address then the
 * REPLY-TO: address. However, it's not until a valid site command is
 * parsed do we know the mailid of the 'defined' admin.
 *
 * So, the net effect is that to earlier you put in logging calls the
 * more change you have of irritating the person on the end of the
 * admin address un-necessarily ;>
 */

int main(int argc, char **argv)
{
    int changed;

    umask(UMASK);

    parse_options(argc, argv);	/* Decode our command line options */

    quit_flag = FALSE;
    parse_headers();		/* Crack the inbound mail headers */
    changed = parse_commands();	/* Process the body of the mail */

    if (!seen_site)
	gupout(0, "No 'site' command found");

    if (changed) {		/* Only sort & prune if changed */
	group_list = sort_groups(group_list);
	if (do_prune) {
	    logit(L_MAIL, "");
	    logit(L_BOTH | L_TIMESTAMP, "Checking new group list");
	    prune(active_list, group_list);
	}
	logit(L_BOTH | L_TIMESTAMP, "Writing updated group list");
	write_groups();
    }
    gupout(0, NULL);
    /* NOTREACHED */
}

/* Parse the command line options */
static void parse_options(int argc, char **argv)
{
    int cc;

    active_path = ACTIVE_PATH;
    newsgroups_path = NEWSGROUPS_PATH;
    mail_fp = mail_open(FALSE, BACKSTOP_MAILID, MAIL_COMMAND, NULL);
    maxgroups = 32767;

    while ((cc = getopt(argc, argv, "a:d:hs:l:n:m:M:PvG:")) != -1) {
	switch (cc) {
	case 'a':
	    active_path = optarg;
	    break;
	case 'd':
	    if (optarg)
		if (chdir(optarg))
		    gupout(1, "Could not chdir to %s", optarg);
	    break;
	case 'h':
	    fprintf(stderr, usage, progname);
	    exit(0);
	case 's':
	    site_directory = optarg;
	    break;
	case 'l':
	    if (optarg) {
		if (!(log_fp = fopen(optarg, "a")))
		    gupout(1, "%s: Could not open log file: %s\n",
			    progname, optarg);
		if (dup2(fileno(log_fp), fileno(stderr)) < 0)
		    gupout(1, "dup2() failed");
	    }
	    break;
	case 'm':
	    mail_fp = mail_open(FALSE, NULL, NULL, optarg);
	    break;
	case 'M':
	    mail_fp = mail_open(FALSE, NULL, optarg, NULL);
	    break;
	case 'P':
	    do_prune = FALSE;
	    break;
	case 'n':
	    newsgroups_path = optarg;
	    break;
	case 'v':
	    print_version();
	    exit(0);
	case 'G':
	    maxgroups = atoi(optarg);
	    break;
	default:
	    gupout(1, "Installation problem. Invalid option used");
	}
    }


    /* Must have an active file that can be opened */
    if (!active_path || !*active_path) {
	fprintf(stderr, "%s: Must nominate the active file with -a\n",
		progname);
	fprintf(stderr, usage, progname);
	gupout(1, NULL);
    }
}

static void print_version(void)
{
    fprintf(stderr, "%s: Version: %s.\n", progname, VERSION);
    fprintf(stderr, "CONFIG_FILENAME:  '%s'\n", CONFIG_FILENAME);
    fprintf(stderr, "ACTIVE_PATH:      '%s'\n", ACTIVE_PATH);
    fprintf(stderr, "NEWSGROUPS_PATH:  '%s'\n", NEWSGROUPS_PATH);
    fprintf(stderr, "MAIL_COMMAND:     '%s'\n", MAIL_COMMAND);
    fprintf(stderr, "BACKSTOP_MAILID:  '%s'\n", BACKSTOP_MAILID);
    fprintf(stderr, "UMASK:            %03o\n", UMASK);
#ifdef USE_FLOCK
    fprintf(stderr, "LOCKING uses:     flock(2)\n");
#endif
#ifdef USE_DOTLOCK
    fprintf(stderr, "LOCKING uses:     dot locking\n");
#endif
}

/*
 * Parse the mail headers and note the interesting ones. Note that we
 * stash all of the interesting headers for later logging and use.
 * Note that this routine avoids calling logit() because we are so
 * early on in the process that our mailid is bound to be the backstop
 * mailid.
 */
static void parse_headers(void)
{
    while (getoneline(stdin)) {
	if (!curr_line[0])
	    return;

	if (strncasecmp(curr_line, "FROM:", 5) == 0) {
	    char buf1[LG_SIZE], buf2[LG_SIZE];

	    CrackFrom(buf1, buf2, curr_line + 5);
	    h_from = xstrdup(buf1);
	    if (!h_reply_to)
		mail_fp = mail_open(FALSE, h_from, NULL, NULL);
	    continue;
	}
	if (strncasecmp(curr_line, "REPLY-TO:", 9) == 0) {
	    char buf1[LG_SIZE], buf2[LG_SIZE];

	    CrackFrom(buf1, buf2, curr_line + 9);
	    h_reply_to = xstrdup(buf1);
	    mail_fp = mail_open(FALSE, h_reply_to, NULL, NULL);
	    continue;
	}
	if (strncasecmp(curr_line, "DATE:", 5) == 0) {
	    h_date = xstrdup(curr_line);
	    continue;
	}
	if (strncasecmp(curr_line, "SUBJECT:", 8) == 0) {
	    h_subject = xstrdup(curr_line);
	    continue;
	}
	if (strncasecmp(curr_line, "MESSAGE-ID:", 11) == 0) {
	    h_message_id = xstrdup(curr_line);
	    continue;
	}
	if (strncasecmp(curr_line, "RETURN-PATH:", 12) == 0) {
	    h_return_path = xstrdup(curr_line);
	    continue;
	}
    }
}

/* Process the commands in the body of the mail */

static int parse_commands(void)
{
    const char *command;
    COMMAND *cmdp;
    const char *tokens[10];
    int tcount;
    int changed;

    changed = FALSE;

    while (!quit_flag && getoneline(stdin)) {
    /* It's safe to log commands after we've seen the site as mail is right */
	if (!*curr_line)
	    continue;

	if (seen_site) {
	    logit(L_MAIL, "");
	    logit(L_BOTH, "%s", curr_line);
	}
	tcount = tokenize(curr_line, tokens, 10);
	if (tcount == 0)
	    continue;

	/* Search for a matching command */
	command = tokens[0];
	for (cmdp = C_list; cmdp->name; cmdp++)
	    if (command_cmp(command, cmdp->name) == 0)
		break;

	/* Find it? */
	if (!cmdp->name) {	/* No */
	    if (!seen_site)	/* What a hack - sigh */
		logit(L_BOTH, "\n%s %s %s", tokens[0],tokens[1],tokens[2]);
	    logit(L_BOTH, "\nERROR: Unrecognized command '%s' - try 'help'\n"
		    "NOTE: No changes made to subscription list.", command);
	    gupout(0, NULL);
	}

	/* Does this command require a preceding site command? */
	if (cmdp->needs_site && !seen_site)
	    gupout(0, "'site' command must precede the '%s' command",
		    cmdp->name);

	/* Only check # of args if defined as > 0 */
	if ((cmdp->args > 0) && (tcount != cmdp->args)) {
	    logit(L_BOTH, "WARNING: %s requires %d parameters, not %d\n",
		    command, cmdp->args, tcount);
	    continue;
	}

	/* Dispatch to handler */
	if (cmdp->handler(tokens + 1))
	    changed = TRUE;

	/* Auto quit? */
	if (cmdp->sets_quit)
	    quit_flag = TRUE;
    }

    return changed;
}

/*      site    sitename        passwd */
static int site(const char **tokens)
{
    FILE *fp;
    int fd;
    const char *hf_tokens[4];
    int found_site;
    int tcount;
    LIST *exclusion_list = NULL;
    char lbuf[MAX_LINE_SIZE];
    char *cryptstr;

    if (seen_site)
	gupout(0, "Multiple 'site' commands not allowed");

    /* Search the config file for a valid sitename */
    if (!(fp = fopen(CONFIG_FILENAME, "r")))
	gupout(1, "Install error. Open of '%s' failed", CONFIG_FILENAME);

    /* Search for matching site */
    found_site = FALSE;
    while (fgets(lbuf, sizeof(lbuf), fp)) {
	tcount = tokenize(lbuf, hf_tokens, 4);
	if (tcount == 0)
	    continue;
	if (*hf_tokens[0] == '#')
	    continue;		/* skip comments -- Md */
	if (tcount != 3)
	    gupout(1, "%s format incorrect for %s. Need 3 tokens",
		    CONFIG_FILENAME, tokens[0]);
	if (strcasecmp(tokens[0], hf_tokens[0]) == 0) {
	    found_site = TRUE;
	    break;
	}
    }

    fclose(fp);

    if (!found_site)
	gupout(0, "Unknown site '%s'", tokens[0]);

/*
 * Ok. We now have the *correct* admin mail ID. Start the mail output.
 * We set up even before completing the site command so that password
 * cracking attempts go to the site admin, not the sender.
 */
    mail_fp = mail_open(TRUE, hf_tokens[2], NULL, NULL);

/*
 * Right.  Now mail_fp is guaranteed to be open, and stdout points there as
 * well.  Let's catch up on all that good log stuff we want to tell them.
 */
    logit(L_LOG, "");
    logit(L_BOTH | L_TIMESTAMP, "BEGIN: %s %s\n", progname, VERSION);

    log_mail_headers();

    logit(L_BOTH, "site %s", tokens[0]);

    /* Found the site how does the password look? */
    cryptstr = crypt(tokens[1], hf_tokens[1]);

    if (strcmp(cryptstr, hf_tokens[1]) != 0		/* crypted pwd */
#ifdef CLEARTEXT_PWD
	&& strcmp(tokens[1], hf_tokens[1]) != 0)	/* cleartext pwd */
#endif
	gupout(0, "Password incorrect");

    sitename = xstrdup(hf_tokens[0]);
    seen_site = TRUE;

    /* It's all good, accept this site as the directory name and cd to it */
    while (chdir(site_directory))
	if (errno == ENOENT) {
	    logit(L_LOG, "MKDIR: %s", site_directory);
	    if (mkdir(site_directory, 0777))
		gupout(1, "Install error: could not MKDIR %s (%s)",
			site_directory, strerror(errno));
	} else
	    gupout(1, "Install error: could not CHDIR to %s (%s)",
		    site_directory, strerror(errno));

    while (chdir(sitename))
	if (errno == ENOENT) {
	    logit(L_LOG, "MKDIR: %s", sitename);
	    if (mkdir(sitename, 0777))
		gupout(1, "Install error: could not MKDIR %s (%s)",
			sitename, strerror(errno));
	} else
	    gupout(1, "Install error: could not CHDIR to %s (%s)",
		    sitename, strerror(errno));

    if (!lockit())
	gupout(1, "Failed to take lock in %s", lbuf);
    if ((fd = open(EXCLUSIONS_FILENAME, O_RDONLY)) >= 0) {
	exclusion_list = read_groups(fd, (LIST *) NULL);
	close(fd);
    }
    load_active(exclusion_list);

    /* Open the sites' groups file - it's ok if it's not there btw */
    if ((fd = open(BODY_FILENAME, O_RDONLY)) >= 0) {
	group_list = read_groups(fd, (LIST *) NULL);
	close(fd);
    } else {
	group_list = create_list();	/* we need something... */
	logit(L_LOG, "WARNING: File '%s' not found for %s", BODY_FILENAME,
		sitename);
    }

    return FALSE;
}

static int quit(const char **tokens)
{
    return 0;
}

/* Check and load the active file */
static void load_active(LIST *exclusion_list)
{
    int fd;

    if ((fd = open(active_path, O_RDONLY)) < 0)
	gupout(1, "Could not open active file '%s'", active_path);
    active_list = read_groups(fd, exclusion_list);
    close(fd);

    if (!active_list->length)
	gupout(1, "No groups found in active file '%s'", active_path);
}

/*
 * Writes the new group list back to disk moderately safely.
 */
static void write_groups(void)
{
    FILE *fp;
    GROUP *gp;
    int write_count;

    if (!(fp = fopen(NEWBODY_FILENAME, "w")))
	gupout(1, "Could not open %s for writing", NEWBODY_FILENAME);
    write_count = 0;
    TRAVERSE(group_list, gp) {
	write_count++;
	if (gp->u.not == GUP_EXCLUDE)
	    putc('!', fp);
	else if (gp->u.not == GUP_POISON)
	    putc('@', fp);
	fputs(gp->name, fp);
	putc('\n', fp);
    }

    fclose(fp);

    logit(L_BOTH | L_TIMESTAMP, "%s: %s: group patterns: %d",
	    progname, sitename, write_count);

/*
 * In the directory is:       groups, groups.new and groups.old
 *
 * rename(groups,old)
 * rename(new,groups)
 */

    if ((rename(BODY_FILENAME, OLDBODY_FILENAME) == -1) && (errno != ENOENT))
	gupout(1, "rename(%s,%s) failed for %s. Errno=%d",
		BODY_FILENAME, OLDBODY_FILENAME, sitename, errno);
    if (rename(NEWBODY_FILENAME, BODY_FILENAME) == -1)
	gupout(1, "rename(%s,%s) failed for %s. Errno=%d",
		NEWBODY_FILENAME, BODY_FILENAME, sitename, errno);
}

/* Include a pattern into the site's group list */
static int include(const char **tokens)
{
    return insert_group(GUP_INCLUDE, tokens[0]);
}

static int exclude(const char **tokens)
{
    return insert_group(GUP_EXCLUDE, tokens[0]);
}

static int poison(const char **tokens)
{
    return insert_group(GUP_POISON, tokens[0]);
}

static int insert_group(int not_flag, const char *gname)
{
    GROUP *gp, *new_gp;
    int match_count;
    int column, indent;

    /* Be optimistic, create the group now */
    new_gp = create_group(not_flag, gname);

    /* See if the add pattern matches anything in active */
    fputs("  => ", mail_fp);
    column = indent = 5;

    match_count = 0;
    TRAVERSE(active_list, gp) {
	if (wildmat(gp->name, new_gp->name)) {
	    match_count++;
	    if (match_count <= LOG_MATCH_LIMIT) {
		item_print(indent, &column, gp->name);
		if (match_count == LOG_MATCH_LIMIT)
		    item_print(indent, &column, "...etc...");
	    }
	}
    }

    if (match_count == 0) {
	if (not_flag == GUP_INCLUDE) {
	    destroy_group(new_gp);
	    logit(L_MAIL, "WARNING: %s pattern not in active - ignored", gname);
	    return FALSE;
	} else
	    logit(L_MAIL, "WARNING: %s pattern not in active", gname);
    }
    logit(L_MAIL, "");
    if (match_count > 1)
	logit(L_MAIL, "  [ %d groups ]", match_count);

    /* Link the new add pattern into the end */
    add_group(group_list, new_gp);

    return TRUE;
}

/* Delete matches on the group list, not the active list */
static int delete(const char **tokens)
{
    const char *gname;
    GROUP *gp;
    int match_count;

    gname = tokens[0];

    /* See if the delete pattern matches anything in the current groups */
    match_count = 0;
    TRAVERSE(group_list, gp) {
	if (wildmat(gp->name, gname)) {
	    match_count++;
	    if (match_count <= LOG_MATCH_LIMIT)
		logit(L_MAIL, "\tMatches: %s %s",
			(gp->u.not == GUP_EXCLUDE) ? "exclude" : "include",
			gp->name);
	    if (match_count == LOG_MATCH_LIMIT)
		logit(L_MAIL, "\tMatches: ...etc...");
	    remove_group(group_list, gp);
	}
    }

    logit(L_BOTH, " delete: Match count %d", match_count);
    if (match_count == 0)
	logit(L_MAIL, "WARNING: %s pattern not in %s - ignored",
		gname, BODY_FILENAME);
    return (match_count > 0);
}

static void item_print(int indent, int *column, const char *str)
{
#define MAX_COL	77	/* number of columns in a line, with room for a ", " */
    int len = strlen(str);

    if (*column + len > MAX_COL) {
	int col = indent;
	char ch;

	putc(',', mail_fp);
	putc('\n', mail_fp);

	if (col % 8)
	    ch = ' ';
	else {
	    ch = '\t';
	    col /= 8;
	}
	while (col--)
	    putc(ch, mail_fp);

	*column = indent;
    }
    if (*column != indent) {
	putc(',', mail_fp);
	putc(' ', mail_fp);
    }
    fputs(str, mail_fp);

    *column += len + 2;
}

/* List out the currently selected groups */
static int list(const char **tokens)
{
    GROUP *gp;

    group_list = sort_groups(group_list);
    TRAVERSE(group_list, gp) {
	logit(L_MAIL, "  %s  %s\n",
		(gp->u.not != GUP_INCLUDE) ?
			((gp->u.not == GUP_POISON) ? " poison" : "  exclude")
		: "include",
		gp->name);
    }
    return 0;
}

/* List out the active file entries that match the pattern */
static int newsgroups(const char **tokens)
{
    print_newsgroups(mail_fp, tokens[0]);
    return 0;
}

/*
 * Read the next line into curr_line. Trim trailing whitespace.
 * Return TRUE if a line was read, otherwise return FALSE.
 */
static int getoneline(FILE *fp)
{
    int len;

    if (!fgets(curr_line, sizeof(curr_line) - 1, fp)) {
	quit_flag = TRUE;
	return FALSE;
    }
    curr_line[sizeof(curr_line) - 1] = '\0';

    /* Trim trailing whitespace */
    len = strlen(curr_line);
    while ((len > 0) && isspace(curr_line[len - 1]))
	len--;
    curr_line[len] = '\0';

    return TRUE;
}

/*
 * A simple whitespace tokenizer - no continutation, no quotes, no
 * nuthin' Does accept comments '#' and empty lines.
 */
static int tokenize(char *cp, const char **tokens, int max_tokens)
{
    int tix;
    char *com_cp;

    /* Find comment */
    for (com_cp = cp; *com_cp; com_cp++)
	if (*com_cp == '#') {
	    *com_cp = '\0';
	    break;
	}

    /* Why don't I use scanf? Good question... */
    for (tix = 0; tix < max_tokens; tix++)
	tokens[tix] = "";

    for (tix = 0; tix < max_tokens; tix++) {
	while (*cp && isspace(*cp))
	    cp++;		/* Skip leading whitespace */
	tokens[tix] = cp;

	while (*cp && !isspace(*cp))
	    cp++;		/* Skip data */

	if (cp == tokens[tix])
	    break;		/* End of data stream */
	if (*cp)
	    *cp++ = '\0';
    }

    return tix;
}

static int command_cmp(const char *str, const char *cmd)
{
    int base = 0;		/* have we matched the base of the cmd? */

    while (*str && *cmd) {
	if (tolower(*str) != *cmd)
	    return TRUE;	/* match failed */
	str++;
	cmd++;
	if (*cmd == '*') {
	    base = TRUE;
	    cmd++;
	}
    }
    return !(!*str && (base || !*cmd));
}

void gupout(int val, const char *out_msg, ...)
{
    va_list ap;

    unlockit();			/* release our lock */

    if (out_msg) {
	va_start(ap, out_msg);
	vlogit(L_BOTH, "ERROR", out_msg, ap);
	va_end(ap);
    }

    log_mail_headers();
    logit(L_BOTH, "");
    logit(L_BOTH | L_TIMESTAMP, "END: %s", progname);

    if (log_fp)
	fclose(log_fp);
    if (mail_fp)
	mail_close(mail_fp);

    /* keep the MTA happy - "don't worry, it's all under control  - honest!" */
/*   exit(val) */

    /* but only if it's not a system error -- Md*/
    if (val > 0)
	exit(75);		/* EX_TEMPFAIL */
    else
	exit(0);
}

/*
 * The mail headers are stashed up by the parse because we don't really
 * want to log them until we absolutely have to. This routine logs them
 * if they haven't already been logged. I don't know that freeing is
 * doing anything much for anyone, but...
 */
static void log_mail_headers(void)
{
    int log_count = 0;

    if (h_from) {
	logit(L_BOTH, "Mail Header: FROM: %s", h_from);
	free(h_from);
	h_from = NULL;
	log_count++;
    }
    if (h_reply_to) {
	logit(L_BOTH, "Mail Header: REPLY-TO: %s", h_reply_to);
	free(h_reply_to);
	h_reply_to = NULL;
	log_count++;
    }
    if (h_subject) {
	logit(L_BOTH, "Mail Header: %s", h_subject);
	free(h_subject);
	h_subject = NULL;
	log_count++;
    }
    if (h_date) {
	logit(L_BOTH, "Mail Header: %s", h_date);
	free(h_date);
	h_date = NULL;
	log_count++;
    }
    if (h_message_id) {
	logit(L_BOTH, "Mail Header: %s", h_message_id);
	free(h_message_id);
	h_message_id = NULL;
	log_count++;
    }
    if (h_return_path) {
	logit(L_BOTH, "Mail Header: %s", h_return_path);
	free(h_return_path);
	h_return_path = NULL;
	log_count++;
    }
    if (log_count)
	logit(L_BOTH, "");
}
